src/python/WMCore/Services/DBS/DBS3Reader.py

```diff
@@ -934,27 +924,51 @@ def fixMissingParentageDatasets(self, childDataset, insertFlag=True):
         :param childDataset: child dataset need to set the parentage correctly.
         :return: blocks which failed to insert parentage. for retry
         """
-        pDatasets = self.listDatasetParents(childDataset)
-        self.logger.info("Parent datasets for %s are: %s", childDataset, pDatasets)
-        # print("parent datasets %s\n" % pDatasets)
-        # pDatasets format is
-        # [{'this_dataset': '/SingleMuon/Run2016D-03Feb2017-v1/MINIAOD', 'parent_dataset_id': 13265209, 'parent_dataset': '/SingleMuon/Run2016D-23Sep2016-v1/AOD'}]
-        if not pDatasets:
+        parentDatasets = self.listDatasetParents(childDataset)
+        self.logger.info("Parent datasets for %s are: %s", childDataset, parentDatasets)
+        # parentDatasets format is
+        # [{'this_dataset': '/SingleMuon/Run2016D-03Feb2017-v1/MINIAOD',
+        #   'parent_dataset_id': 13265209,
+        #   'parent_dataset': '/SingleMuon/Run2016D-23Sep2016-v1/AOD'}]
+        if not parentDatasets:
             self.logger.warning("No parent dataset found for child dataset %s", childDataset)
             return {}
 
-        parentFullInfo = self.getParentDatasetTrio(childDataset)
+        parentFlatData = self.getParentDatasetTrio(childDataset)
+
         blocks = self.listBlocksWithNoParents(childDataset)
         failedBlocks = []
         self.logger.info("Found %d blocks without parentage information", len(blocks))
         for blockName in blocks:
             try:
                 self.logger.info("Fixing parentage for block: %s", blockName)
-                numFiles = self.findAndInsertMissingParentage(blockName, parentFullInfo, insertFlag=insertFlag)
-                self.logger.debug("%s file parentage added for block %s", numFiles, blockName)
+                childFlatData = self.getChildBlockTrio(blockName)
```

Comment:
If I read this code correctly the complexity of this loop will be O(n^4) and the code will be extremely slow when number of processed blocks will go above 10 or similar scale. Instead I suggest to take this codebase and create a single function which will only take block name and perform the rest. Then you may use asyncio to perform parallel execution of this task for all blocks.

Reply:
Hi @vkuznet where did you see a O(n^4) complexity here? 
I can suspect where, but it is not the case if you define N as number of blocks as you state few lines bellow. 

If you refer to the two additional nested `for` cycles which are coming from inside the   `self.getChildBlockTrio(blockName)` call ... well lets look at them one at a time:

https://github.com/dmwm/WMCore/pull/11520/files#diff-0f2f04c342289791ed9a6b465578ccdf229ad9e28bdb98a93651e599ae5b0c1cR1018-R1021

The first one just goes through the fileids, while the other one simply iterates through the run/lumi pairs associate with each file. So no way to avoid that.

As of the other two `for` cycles that are bellow and try to match the parentage information between the parent dataset and the current child block: 
 * Firstly they are not nested 
 * Secondly if you are concerned about the set operations: The complexity of `dict.keys()` for python 3  is O(1) and the set's diff (`-`) and  intersection(`&`) operations are not of complexity of O(n) but of O(`size of the smaller set) to the worst: https://wiki.python.org/moin/TimeComplexity. So in our case  the size of the block. 

So if you define  N (in the worst case) as number of `run/lumi` pairs well then ... the current situation is simply unavoidable. 

And honestly wrapping the `for` in a single function/method won't solve the issue with code complexity, no matter how you do that, because in order  to "flatten" a triple nested data structure (as the one returned from DBS) one can get with at least O(n^3) complexity (unless trying to do some algorithmic optimization etc... which for the current use case would be an overkill). 

So instead of trying to flatten the data at WMCore then I'd say we  may try to fix the API on the DBS side.

Reply:
You are right the complexity is O(n^3) and it comes from this parts:
```
        for blockName in blocks:
            try:
                self.logger.info("Fixing parentage for block: %s", blockName)
                childFlatData = self.getChildBlockTrio(blockName)
...
    def getChildBlockTrio(self, childBlock):
         ...
        for fileId in childBlockInfo[0]:
            for runLumiPair in childBlockInfo[0][fileId]:
```
So we have three nested loops, one for block names, then API itself make to loop.

That's why I said that it is better to discuss API and data structure you need and implement it on DBS server side instead of having such complex codebase in a client.

Reply:
We should be moving with what  we have so far. And with the current PR  things are going to get better now.